frontline-functions-quickstart/functions/callbacks/crm.protected.js

```diff
@@ -0,0 +1,83 @@
+/* eslint-disable camelcase */
+const handleGetCustomersListCallback = async (context, event) => {
+  const customersFile = Runtime.getAssets()['/providers/customers.js'].path;
+  const { getCustomersList } = require(customersFile);
+
+  console.log('Getting Customers list');
+
+  const workerIdentity = event.Worker;
+  const pageSize = event.PageSize;
+  const anchor = event.Anchor || 0;
+
+  console.log('worker identity: ', workerIdentity);
+
+  // Fetch Customers list based on information about a worker, that requested it
+  const customersList = await getCustomersList(
+    context,
+    workerIdentity,
+    pageSize,
+    anchor
+  );
+
+  // Respond with Customers object
+  return {
+    objects: {
+      customers: customersList,
+    },
+  };
+};
+
+const handleGetCustomerDetailsByCustomerIdCallback = async (event, context) => {
+  const customersFile = Runtime.getAssets()['/providers/customers.js'].path;
+  const { getCustomerById } = require(customersFile);
+  console.log('Getting Customer details: ', event.CustomerId);
+
+  const customerId = event.CustomerId;
+
+  /**
+   *  Fetch Customer Details based on their ID
+   * and information about a worker, that requested that information
+   */
+  const customerDetails = getCustomerById(context, customerId);
+
+  // Respond with Contact object
+  return {
+    objects: {
+      customer: {
+        customer_id: customerDetails.customer_id,
+        display_name: customerDetails.display_name,
+        channels: customerDetails.channels,
+        links: customerDetails.links,
+        avatar: customerDetails.avatar,
+        details: customerDetails.details,
+      },
+    },
+  };
+};
+
+exports.handler = async function (context, event, callback) {
+  const location = event.Location;
+
+  /**
+   * Location helps to determine which information was requested.
+   * CRM callback is a general purpose tool and might be used to fetch different kind of information
+   */
+  switch (location) {
+    case 'GetCustomerDetailsByCustomerId': {
+      const resp = await handleGetCustomerDetailsByCustomerIdCallback(
+        event,
+        context
+      );
+      return callback(null, resp);
+    }
+    case 'GetCustomersList': {
+      const resp = await handleGetCustomersListCallback(context, event);
+      return callback(null, resp);
+    }
+
+    default: {
+      console.log('Unknown location: ', location);
+      return callback(422, `Unknown location:  ${location}`);
```

Comment:
This unfortunately doesn't work (applies throughout). If you pass anything but `null` as the first argument the HTTP status code will always be `500` and the response body will be the value of the first argument as string. So if you want to return an HTTP status code of 422 you'd have to do this:

```suggestion
      const response = new Twilio.Response({
        statusCode: 422,
        body: `Unknown location: ${location}`
      });
      return callback(null, response);
```

Reply:
@dkundel I made this change but it passes the status as `200` instead. Looks like `Twilio.Response` doesn't pass the `422` through. I'm trying to find out more about how it works, but in the meantime, is the use of `Twilio.Response` required or can I send a plain object with the values?

Reply:
All set, I set the `setStatusCode` and `setBody` methods from `Twilio.Response` and it works. 
More info [here](https://www.twilio.com/docs/runtime/functions/invocation#twilio-response-methods).